### DISCUSSION TITLE
Clustering the nodes of the Cora dataset

### DISCUSSION THREAD

**Discussion by OliverSummers** [pyg-team/pytorch_geometric#2819] [Q&A] [answered]
I want to cluster the nodes of the Cora dataset based on the in-degree value of the nodes so that each cluster only contains the nodes with the same in-degree value. I know how to get the counts:

from torch_geometric.datasets import Planetoid
from torch_geometric.utils import degree

dataset = Planetoid('./data','CORA')
d = dataset[0]
n = d.num_nodes
indegree = degree(d.edge_index[1], n, dtype=torch.long)
counts = torch.bincount(indegree)

But since I don't access the index value of the nodes, I don't know how to place each node in which cluster?

**Comment by rusty1s** [comment, marked as answer]
How about using `argsort()` to re-sort your data so that nodes with the same node degree have a contiguous memory layout? For example, for re-permuting the node features, you can do:
```python
perm = indegree.argsort()
data.x = data.x[perm]
```
For re-permuting edge indices, you can do:
```python
data.edge_index = perm[data.edge_index]
```

**Reply by OliverSummers** [reply]
Thank you for your useful suggestion. But still, I am not sure about how to place each node in a cluster since I don't access the index value of the nodes, any suggestions?

**Reply by rusty1s** [reply]
I see. Do you want to build individual subgraphs out of nodes with the same degree? In case yes, you can use `torch_geometric.utils.subgraph(indegree = x, ...)` to achieve that, where `x` is replaced by an integer.

If you solely want to group all node features, you can split the node features via `torch.split` based on your `bincount` vector after permutation.

**Reply by OliverSummers** [reply]
Thank you very much.
